core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -330,6 +344,25 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
       completed
     }
 
+    // traverse the list and try to complete some watched elements
+    def tryCompleteFirstWatched(): Boolean = {
```

Comment:
Actually the reason I wanted to complete one at a time is that, as @apurvam also mentioned, upon completing each delayed operation we are calling another `completeWatched` again on the same list, so let's say if we have three operations in the list `(1, 2, 3)`, then calling the `completeWatched` call will trigger the `tryComplete` on the order of:

```
1, 2, 3 (as a result of completing 1), 2, 3 (as a result of completing 2), 3,
```

And even worse, each `tryCompleteElseWatched` call will trigger the `tryUpdate` again; and when the `tryComplete` on some of the operations on the list has failed it will add the operation again on the tail of the list again and hence the `completeWatched` will never end.

Thinking about it a bit more, I think it would be a better fit for a blocking queue than use this purgatory even. Thoughts?